src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +872,136 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+    });
+}
+
+ss::future<> ntp_archiver::housekeeping() {
+    if (housekeeping_can_continue()) {
+        co_await apply_retention();
+        co_await garbage_collect();
+    }
+}
+
+ss::future<> ntp_archiver::apply_retention() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    auto log = _partition_manager.log(_ntp);
+    if (!log) {
+        co_return;
+    }
+
+    auto retention_calculator = retention_calculator::factory(
+      manifest(), _partition->get_ntp_config(), *log);
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        vlog(
+          _rtclog.debug,
+          "{} Advancing start offset to {} satisfy retention policy",
+          retention_calculator->strategy_name(),
+          *next_start_offset);
+
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
+          *next_start_offset, deadline, _as);
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.warn,
+              "Failed to update archival metadata STM start offest according "
+              "to retention policy: {}",
+              error);
+        }
+    } else {
+        vlog(
+          _rtclog.debug,
+          "{} Retention policies are already met.",
+          retention_calculator->strategy_name());
+    }
+}
+
+ss::future<> ntp_archiver::garbage_collect() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    std::vector<remote_segment_path> to_remove;
+    auto current_start_offset = manifest().get_start_offset();
+
+    for (const auto& [key, value] : manifest()) {
+        if (value.committed_offset < current_start_offset) {
+            to_remove.emplace_back(
+              manifest().generate_remote_segment_path(_ntp, key, value));
+        }
+    }
+
+    auto replaced = manifest().replaced_segments();
+    for (const auto& meta : replaced) {
+        const auto segment_name = cloud_storage::generate_local_segment_name(
+          meta.base_offset, meta.segment_term);
+        to_remove.emplace_back(cloud_storage::generate_remote_segment_path(
+          _ntp, meta.ntp_revision, segment_name, meta.archiver_term));
+    }
+
+    // TODO(vlad): Maybe make this loop concurrent.
+    auto deletions_failed = false;
+    for (const auto& path : to_remove) {
+        retry_chain_node fib(
+          _metadata_sync_timeout, _cloud_storage_initial_backoff, &_rtcnode);
+        retry_chain_logger ctxlog(archival_log, fib, _ntp.path());
+
+        // If deletion fails and the subsequent command to cleanup metadata
+        // succeeds, the segment will become orphaned.
+        auto res = co_await _remote.delete_object(
```

Comment:
There should be some audit trails for segment deletion on info level. If we will ever have a data loss in the cloud storage the retention and GC will be a prime suspect. We will need a way to tell when something is deleted (or most importantly not deleted) by the retention without bumping logs to debug or trace.

Reply:
+1, these are very verbose messages but worth having.  It will be the same situation as local storage logs, where by default we log things like starting a new segment, but for high scale systems we provide advice (https://vectorizedio.atlassian.net/wiki/spaces/CORE/pages/99844097/Handling+Large+Partition+Counts+in+Redpanda+22.2#Logging) on which subsystems to tune down to debug level to manage log volume.

I will update that scale wiki page to mention the cloud storage loggers too.

Reply:
Makes sense to me. I'll add some logging that includes the segment names at info level. Perhaps in the future we can have some in memory store that can be queried via `rpk audit` (or something similar).

Reply:
Done. All segment deletions are logged at the info level with the remote path.